ccid driver

hrosenfeld opened the pull request:
This is Roberts ccid driver plus a load of cleanups, a few small fixes, and the necessary plumbing to build and ship it in the platform. As a generic ccid driver this is incomplete, but it appears to be working well enough with YubiKeys to be useful.

The two individual commits are also available as webrevs:
http://us-east.manta.joyent.com/hrosenfeld/public/ccid-orig/index.html
http://us-east.manta.joyent.com/hrosenfeld/public/ccid-dev/index.html

jlevon commented:
Can you also do a webrev of master against 2f935e4  please?

hrosenfeld commented:
Full webrev: http://us-east.manta.joyent.com/hrosenfeld/public/ccid/index.html

jlevon commented:
There are lot of unfinished bits here including large numbers of XXX from rm's original changes. Even so far as comments that end in the middle of a sentence. It seems like these should all be resolved before anybody digs further into review.

I understand that some features etc. aren't going to be supported at least initially. In general this seems fine, but half-written code is not something that should really make the gate. Instead, there should be a comment section on what isn't yet implemented (or perhaps bugs filed), and make sure that the code *as is* still makes sense.

As I understand it, this only works for some small subset of CCIDs like certain Yubikeys.
But right now, it will attach to any device matching the class right? What's the plan to fix this?

So I don't have to mention it everywhere: please excise all lint-related changes in Makefiles, headers, C files ("ARGSUSED") etc. No lint should be in new code.

hrosenfeld commented on usr/src/uts/common/io/usb/clients/ccid/ccid.c:
```diff
@@ -0,0 +1,5207 @@
+} ccid_minor_flags_t;
+
+typedef struct ccid_minor {
+	ccid_minor_idx_t	cm_idx;		/* WO */ /* XXX: Whats 'WO'? */
```
Thanks!

pwinder commented:
I have only been through the 2 files below ….

There are a fair few XXX comments - some sound problematic. Can they be cleaned up, if some are known issues could they be documented in bug reports?

There is also a liberal use of VERIFY(). I assume this is a hang over from development. Should these rather not be ASSERT(), so only enabled on DEBUG?

Most of the comments are nits about style. It is all new code, so it would be good to maintain consistent style!

I noticed the ccid_mutex is occasionally dropped temporarily and then re-acquired. Are these unlock/lock safe? They won’t cause bad interaction between parallel execution paths?


common/io/usb/clients/ccid/ccid.c
=========================
Line 978:
Just to be consistent make it:
if ((cmp->cm_flags & CCID_MINOR_F_TXN_RESET) != 0) {

Line 995 and 999:
Similar to above, explicitly compare against != 0

Line 1149:
Both src and dst of bcopy are uint8_t. Can’t you just use a straightforward assignment?
val = cch->ch_param2

Line 1170:
Again for consistency:
if ((cc->cc_flags & CCID_COMMAND_F_USER) != 0)

Line 1373:
s/wnat/want/

Line 1511:
For consistency:
if ((ccid->ccid_flags & CCID_F_DETACHING) != 0)

Lines 1969 and 1970:
Lines 2056 and 2057:
Lines 2113 and 2114:
Can be replaced with:
goto done

Functions ccid_worker(), ccid_worker_request(0, ccid_intr_restart_timeout()
There are a lot of if conditional expression with binary ops similar to those I’ve highlighted above. Can they all be changed to explicitly compare the result of the binary op against 0?

Line 2929:
Add ‘!’ As first character of message.

Line 3132:
s/faild/failed/ and prefix message with ‘!’

Line 3125:
prefix message with ‘!’. Most of the messages are the beginning are prefixed with ‘!”, and it seems a lot from this point on are not.
Maybe do a global search and make them all consistent (with a ‘!’ prefix).

Line 3451:
Make the format a %u imc the snprintf() as the variable is unsigned.

Line 3467:
Add a blank line after the variable declaration.

Function ccid_cleanup():
There a a lot of “if (flag & SOME_VALUE)” please change to “if (flag & SOME_VALUE) != 0)

Line 3929:
Same comment about the if conditional - explicitly compare against 0.

Line 4143:
Because of the return in the “true” path of the the preceding “if, the “} else if” is not need, just have “if ( … )

ccid_read(), ccid_write(), ccid_ioctl_status(), ccid_ioctl_fionread(), ccid_chpoll()
A few cases of - Same comment about the if conditional - explicitly compare against 0.

common/sys/usb/clients/ccid/ccid.h
===========================
Lines 282 - 298
Align the values assigned to various enums as in the other enum declarations.


- Paul

On 5 Nov 2019, at 17:13, Hans Rosenfeld <notifications@github.com<mailto:notifications@github.com>> wrote:

CAUTION: This email originated from outside of Western Digital. Do not click on links or open attachments unless you recognize the sender and know that the content is safe.


Full webrev: http://us-east.manta.joyent.com/hrosenfeld/public/ccid/index.html

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/joyent/illumos-joyent/pull/199?email_source=notifications&email_token=AARKFBONRBMU2KJRTA3G4ZDQSGST7A5CNFSM4JJEMBB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEDDQJDY#issuecomment-549913743>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AARKFBMLHN5CFIGQZVNWPF3QSGST7ANCNFSM4JJEMBBQ>.

hrosenfeld commented:
I've cleaned up the code and comments, implemented missing features removed two big pieces of code which were incomplete and not needed to support YubiKeys. I've also added a few more test cases.

There are still very few XXXs in the code which I haven't resolved yet, but that will follow shortly. I will file a handful of bugs for some features which would be nice to have eventually, including full TPDU protocol support which I've just ripped out for now.

hrosenfeld commented:
I've cleaned up the code and comments, implemented missing features, and removed big pieces of code which were incomplete and not needed to support YubiKeys. I've also added a few more test cases.

There are still very few XXXs in the code which I haven't resolved yet, but that will follow shortly. I will file a handful of bugs for some features which would be nice to have eventually, including full TPDU protocol support which I've just ripped out for now.

jlevon commented:
Can you re-generate the webrev against master please? (I don't think the other individual webrevs are useful any more).

One comment I had from the previous set: please remove all mentions of "lint" from new makefiles/code.

hrosenfeld commented:
Here's the updated webrev: http://us-east.manta.joyent.com/hrosenfeld/public/ccid/index.html

I thought I removed all traces of lint already, if I missed some please let me know.

jlevon commented on usr/src/uts/common/io/usb/clients/ccid/ccid.c:
```diff
@@ -0,0 +1,4426 @@
+
+		if (change) {
+			ccid_worker_request(ccid);
+		
```
s/CCID/ICC/ I think?

jlevon replied:
Please clarify that CLASS_F_EXT_APDU_XCHG is definied to *also* support short apdu, and that's all we support right now, so we only ever set the callbacks to those methods.

jlevon replied:
Does this refer to also CCID_SLOT_F_NEED_IO_TEARDOWN waiters? Either way a comment would be good, presuming those waiters *will* get woken up somehow.

hrosenfeld commented on usr/src/common/ccid/atr.c:
```diff
@@ -0,0 +1,1606 @@
+	 * protocol represents the preferred protocol.
+	 *
+	 * When Ti is two TA, TB, and TC also represent global properties.
+	 * However, TC only has meaning if the protocol is T=0.
```
I'm not sure what you mean here, but it looks as if a comma is missing.

jlevon replied:
I think this was fixed in one of your later pushes.

hrosenfeld commented on usr/src/lib/libpcsc/common/winscard.h:
```diff
@@ -0,0 +1,139 @@
+#define	_WINSCARD_H
+
+/*
+ * This library provides a compatability interface with programs designed
```
PC/SC lite is just another free implementation of the same Windows interface, so I'm not sure we should link to them. I could link to the Microsoft documentation for winscard.h, but I'm not sure that would be all that useful.

jlevon replied:
It'd still be better than nothing right now?

hrosenfeld commented on usr/src/cmd/ccidadm/ccidadm.c:
```diff
@@ -0,0 +1,821 @@
+static ccidadm_cmdtab_t ccidadm_cmds[] = {
+	{ "list", ccidadm_do_list, ccidadm_list_usage },
+	{ "atr", ccidadm_do_atr, ccidadm_atr_usage },
+	{ "reader", ccidadm_do_reader, ccidadm_reader_usage },
```
Whats wrong with "reader"? I think it's clear enough and quicker to type than "reader-info".

jlevon replied:
What if we want other commands that apply to the reader? "ccidadm reader set" etc?

How about just "info"? If we wanted ICC/slot commands we could add slot-info or -s instead.

hrosenfeld replied:
I'd really like to leave this as is. I think "info" is implied - it's either information about the reader, or the cards ATR. I don't expect that we'll ever extend ccidadm to do much more than what it does today, even if the name may suggest that we do.

The way these readers and cards are treated by the driver doesn't really lend itself very well to doing things that are persistent across closing a device. One exception may be support for changing PINs on cards, but that is IIRC dependent on the higher level protocols that the cards use and I'm not sure that would have a place in ccidadm.

hrosenfeld commented on usr/src/lib/cfgadm_plugins/ccid/common/cfga_ccid.c:
```diff
@@ -0,0 +1,416 @@
+	if (ioctl(fd, UCCID_CMD_ICC_MODIFY, modify) != 0) {
+		int e = errno;
+		(void) close(fd);
+		return (cfga_ccid_error(CFGA_ERROR, errp,
```
Closing will do that automatically.

jlevon replied:
And unlike the case below, we really do want to do a reset?

hrosenfeld replied:
Yes, that seems reasonable if 'modify' failed.

hrosenfeld commented on usr/src/lib/cfgadm_plugins/ccid/common/cfga_ccid.c:
```diff
@@ -0,0 +1,416 @@
+		cld->ap_o_state = CFGA_STAT_UNCONFIGURED;
+	}
+
+	/*
```
Yes. I agree with the comment, we could include additional information in the status ioctl. That would be the relatively simple part of it. What I currently don't know and don't want to get into right now is figuring out how represent that information in the cfgadm output.

I'd like to leave this comment in there, XXX included, and file a bug for this before integration.

jlevon replied:
ok

hrosenfeld commented on usr/src/man/man7d/ccid.7d:
```diff
@@ -0,0 +1,426 @@
+.Nm ccid
+.Nd chip card interface device USB client class driver
+.Sh SYNOPSIS
+.Pa /dev/ccid/
```
I'll turn this into /dev/ccid/ccidX/slotX. There's more explanation of the paths further down in the manpage, I could also just remove it, I guess.

jlevon replied:
Fine, but please put it under a "PATHS" section or some such, if you choose to keep it. Right next to a #include isn't right

hrosenfeld commented on usr/src/uts/common/io/usb/clients/ccid/ccid.c:
```diff
@@ -0,0 +1,4426 @@
+		return (EEXIST);
+	}
+
+	if ((cmp->cm_flags & CCID_MINOR_F_WAITING) != 0) {
```
This can happen if an open minor is shared by threads or processes, the driver will still only have one ccid_minor_t for that and doesn't know or care about the actual threads or processes. If both threads owning a single ccid_minor_t attempt to get exclusive access at the same time, one of them will get EINPROGRESS.

I think it would be preferable if we didn't have to worry about this at all and just disallow such use cases. This would be a nice case for O_CLOFORK, too.

jlevon replied:
Couldn't we just get rid of the flag and allow threads to queue up on the CV? Losers might have to re-add 'cmp' to the list again but that seems OK? Is this for when we're trying to tear a slot down maybe?

Alternatively, if we really want to make it impossible for threads to wait on a txn like this, we should at least clearly comment this here - and I can't remember what the manpage says but it should do so there if not.

What do you think?

hrosenfeld replied:
We already handle multiple threads waiting for exclusive access if they all have their own cloned minor. I'm not sure it's justified to complicate this driver more by allowing multiple threads to wait on each other on the same cloned minor. It just seems as a rather obscure use case, and I fail to see what benefit it would give to an application.

I'll check the manpage to see if there's a need to explain this behavior better.

hrosenfeld commented on usr/src/uts/common/io/usb/clients/ccid/ccid.c:
```diff
@@ -0,0 +1,4426 @@
+
+		if (change) {
+			ccid_worker_request(ccid);
+		
```
I removed this. This isn't actually correct, as we still want to attach this reader even if we currently can't do I/O with it. In that case we can still get the ATR and the reader capabilities which could be useful, and is also documented behavior elsewhere in this driver.

hrosenfeld commented on usr/src/uts/common/io/usb/clients/ccid/ccid.c:
```diff
@@ -0,0 +1,4426 @@
+
+		if (change) {
+			ccid_worker_request(ccid);
+		
```
Wouldn't that just add confusion?

The comment states that we have to wake up anyone waiting ccid_read() for I/O to finish, and thats what the code does. Yes, there may be others waiting on cm_iowait_cv for ccid_read() to finish to release their exclusive hold, and they'll be woken up by ccid_read(), but I'm not sure why that would be relevant here?

jlevon replied:
Taking a step back, I think I'm really asking for a description of how a disconnect affects in-progress operations and the locking/wakeup/cleanup involved.

hrosenfeld replied:
I've extended to comment below to mention that a disconnect results in the bulk xfer exception callback to be called when a disconnect occurs. That will clean it up by going through ccid_command_transport_error(). The locking/wakeup is the same as in any other command completion then.

hrosenfeld commented on usr/src/uts/common/io/usb/clients/ccid/ccid.c:
```diff
@@ -0,0 +1,4426 @@
+
+		if (change) {
+			ccid_worker_request(ccid);
+		
```
I can see the reasoning behind that, but then it would really have to be a VERIFY0(). Do you think it's worth doing that all over the driver?

jlevon replied:
You don't need it at all here. By definition that's what ccid_minor_find_user() guarantees, and the correct place for an assert. I think (?) it's the only place that does this double-checking.

hrosenfeld commented on usr/src/uts/common/io/usb/clients/ccid/ccid.c:
```diff
@@ -0,0 +1,4426 @@
+
+		if (change) {
+			ccid_worker_request(ccid);
+		
```
I'm not sure EBUSY would make sense. If a user doesn't have an exclusive hold, they're not allowed to poll just as they're not allowed to read or write. The chpoll(9e) manpage doesn't mention any restrictions on the errno we are allowed to return, and poll(2) does list a few but I'm not sure that list is exhaustive. The only other errno that would possibly make sense if we were to restrict ourselves to what poll(2) states would be EINVAL, as the minor is clearly in an invalid state for polling when not held exclusively. But that doesn't match at all what poll(2) states about the reasons for EINVAL.

jlevon replied:
I disagree. Another user has an exclusive hold and is presumably
therefore using the slot. When it's done it will become available again.
EBUSY is exactly the right return here. There is no permissions involved
here which is what EACCESS is for

hrosenfeld replied:
All this checks is whether the minor which is polled has an exclusive hold. If it doesn't that doesn't mean anything about whether another minor may have an exclusive hold or not, so EBUSY just isn't right as we just don't know whether it's busy.

Perhaps EINVAL is the better error code here, as a minor that hasn't a exclusive hold is in an invalid state for polling. Or perhaps this should just return 0 and indicate neither POLLIN nor POLLOUT as there's just nothing going on currently?

jlevon replied:
I had the test inverted in my head, so you're right, EBUSY isn't correct; thanks. I think EINVAL is the best option here.

hrosenfeld replied:
Thinking more about this over the past day, I decided to return no error and just return no events if not in a transaction. The main reason is that with no transaction you can't write or read, and thats what no events indicate. Also EINVAL is documented in poll(2) to indicate a handful of special cases that don't apply here.

jlevon replied:
OK, but please make sure this is documented clearly.

jlevon commented:
On Mon, Dec 16, 2019 at 10:55:24AM -0800, Hans Rosenfeld wrote:

> I'm not sure EBUSY would make sense.

I disagree. Another user has an exclusive hold and is presumably
therefore using the slot. When it's done it will become available again.
EBUSY is exactly the right return here. There is no permissions involved
here which is what EACCESS is for

hrosenfeld commented on usr/src/cmd/ccidadm/ccidadm.c:
```diff
@@ -0,0 +1,821 @@
+
+	/*
+	 * For an absolute path, just try the base path. If it doesn't match our
+	 * prefix, simulate ENOENT.
```
I removed this special case. If ccidadm gets an absolute path, it'll just open it regardless of whether it's in CCID_ROOT or not.

hrosenfeld commented on usr/src/man/man7d/ccid.7d:
```diff
@@ -0,0 +1,426 @@
+.\"
+.\" This file and its contents are supplied under the terms of the
```
Is that  still the case with the changed chpoll code?

jlevon replied:
I think only if we have additional error semantics than that provided by poll(2)

hrosenfeld commented on usr/src/test/os-tests/tests/Makefile:
```diff
@@ -29,6 +29,7 @@ SUBDIRS =       \
 		stress \
 		timer \
 		tmpfs \
+		uccid \
```
I added atrparse to default.run and added a comment there about the hardware requirement for the other uccid tests.

hrosenfeld commented:
Updated webrev: http://us-east.manta.joyent.com/hrosenfeld/public/ccid/index.html
Webrev with all changes for John's review comments: http://us-east.manta.joyent.com/hrosenfeld/public/ccid-john/index.html

jlevon commented:
On Thu, Dec 19, 2019 at 09:24:03AM -0800, Hans Rosenfeld wrote:

> Updated webrev: http://us-east.manta.joyent.com/hrosenfeld/public/ccid/index.html
> Webrev with all changes for John's review comments: http://us-east.manta.joyent.com/hrosenfeld/public/ccid-john/index.html

Cc'ed you directly as my last attempt to reply didn't show up on github.

Thanks for doing these changes; I just have a few final nits. I can't bear to
attempt to brave the github interface again for them.

Please make sure to file bugs on the remaining XXX we discussed as well
as a way to report "unsupported card" status in ccidadm.

http://us-east.manta.joyent.com/hrosenfeld/public/ccid-john/usr/src/cmd/ccidadm/ccidadm.c.wdiff.html

:73 - you still have the old comment here

http://us-east.manta.joyent.com/hrosenfeld/public/ccid-john/usr/src/lib/libpcsc/common/libpcsc.c.wdiff.html

:304 - this should be LONG - yeah, it amounts to the same thing, but
given the weirdo types we have for this library I'd rather by consistent

:341 - this was there already, but should be LONG too

:504 ditto

http://us-east.manta.joyent.com/hrosenfeld/public/ccid-john/usr/src/uts/common/sys/usb/clients/ccid/ccid.h.wdiff.html

:234 - this wasn't fixed, still need s/Clock Frequency/ClockFrequency/

http://us-east.manta.joyent.com/hrosenfeld/public/ccid-john/usr/src/test/os-tests/tests/uccid/notxn-poll.c.wdiff.html

:18 it's not an error any more

http://us-east.manta.joyent.com/hrosenfeld/public/ccid-john/usr/src/man/man7d/ccid.7d.wdiff.html

:14 - could do with a date bump

:46 - please clarify short apdu only

:163 - I thought we just returned 0 now as per the test

:180 thanks for adding this, but do we want to generalize the fork()
mention - presumably passing fd's between processes in general (over a
socket, exec, etc.) has the same semantics

http://us-east.manta.joyent.com/hrosenfeld/public/ccid-john/raw_files/new/usr/src/man/man1m/ccidadm.1m.man.html

ccidadm atr -p

 - sorry to notice this so late, but usually "-p" means *parseable*.
   Given it's the default, can we just remove this so if we ever want a
   parsable output other than the hexdump, we can add it?

hrosenfeld commented:
New incremental webrev: http://us-east.manta.joyent.com/hrosenfeld/public/ccid-john-2/index.html

hrosenfeld commented:
Filed bugs for follow-on work:
OS-8081 want support for CCID readers lacking APDU and automatic parameter negotiation
OS-8080 implement UCCID_CMD_ERROR_INFO
OS-8079 uccid(7i) status command should include information about driver/reader state
OS-8078 ATR printing for ccid could decode historical bytes
